posts/2023-06-07-kotlin-ktor-keycloak.adoc

```diff
@@ -0,0 +1,357 @@
+
+[source,kotlin]
+----
+install(Authentication) {
+        jwt("auth-jwt") {
+            realm = "lunarealm"
+            verifier(
+                UrlJwkProvider(URL("http://localhost:8080/realms/lunarealm/protocol/openid-connect/certs")),
+                "http://localhost:8080/realms/lunarealm",
+            ) {
+                withAudience("luna-vacation-bot")
+                withClaim("azp", "luna-vacation-bot")
+            }
+            challenge { _, _ ->
+                call.respond(HttpStatusCode.Unauthorized, "Token is not valid or has expired")
+            }
+            validate { credential ->
+                validateCredential(credential, issuer)
+            }
+        }
+    }
+----
+If the authentication fails it returns the `challenge`, in this case a 401 Unauthorized with an error message.
+
+The `validateCredentials` function is simply validating that the token has not expired, after all other validations have passed.
+
+[source,kotlin]
+----
+fun validateCredential(credential: JWTCredential, issuer: String): JWTPrincipal? {
+    if (credential.expiresAt?.after(Date()) == true
+    ) {
+        return JWTPrincipal(credential.payload)
+    }
+
+    return null
+}
+----
+
+==== Test the API
+We can test the api using curl.
+If we secured the endpoint properly, calling the `imionvacation` endpoint without a token should return an error.
+
+[source,commandline]
+----
+$ curl localhost:4040/imionvacation
+Token is not valid or has expired
+----
+
+As the diagram at the beginning of this post explained, we need to request a token from Keycloak and send it together with the request, and for that we impersonate the `luna-vacation-bot`:
+
+[source,commandline]
+----
+curl -X post 'http://localhost:8080/realms/lunarealm/protocol/openid-connect/token' \
+--header 'Content-Type: application/x-www-form-urlencoded' \
+--data-urlencode 'grant_type=client_credentials' \
+--data-urlencode 'client_id=luna-vacation-bot' \
+--data-urlencode 'client_secret=<replace-by-secret>'
+----
+
+Grab the `access_token` part of the response and add it to the request header as a Bearer token:
+
+[source,commandline]
+----
+$ curl 'localhost:4040/imionvacation' \
+--header 'Authorization: Bearer eyJhbGciOiJ...QBhNiX6w'
+----
+
+And now we get a result:
+
+[source,commandline]
+----
+Today you have to work
+----
+
+
+=== Create the luna-vacation-bot
+
+To create the `luna-vacation-bot` project we are going to use https://www.jetbrains.com/idea/[Intellij] (or your favorite IDE) and create a new Kotlin application, with https://gradle.org/[Gradle] and the latest JVM you have available (I'm using JDK 17). After the project is created you can remove the test folder and the test dependencies.
```

Comment:
```suggestion
To create the `luna-vacation-bot` project we are going to use https://www.jetbrains.com/idea/[Intellij] (or your favorite IDE) and create a new Kotlin application, with https://gradle.org/[Gradle] and the latest JVM you have available (I'm using JDK 17). After creating the project we can remove the test folder and the test dependencies.
```
I don't think saying "we are going to use.... (or your favorite IDE)". Stick to using one pronoun and avoid switching from first-plural to second person.

Maybe instead of saying "the latest JVM you have", provide the minimum JVM version requirement.

Reply:
what do you think of 
```To create the `luna-vacation-bot` project we are going to use https://www.jetbrains.com/idea/[Intellij] and create a new Kotlin application, with https://gradle.org/[Gradle] and JDK 17 (you can use JDK 8 or higher).```